src/doc/types.md

```diff
@@ -0,0 +1,30 @@
+# Types
+
+Certain types within the Tock kernel are chosen very deliberately. This aids
+with clarity on how the kernel works and helps with porting to new platforms.
+
+## Integer Types
+
+Many kernel interfaces use integer types (e.g., `usize` or `u32`). These types
+are often used because the contained value must fit in a hardware register,
+represents a memory address, or is a fixed width in an interface specification.
+The kernel should use these types very deliberately depending on the meaning of
+the underlying value.
+
+- Fixed-width integers (e.g., `u32`, `u8`, etc.): These types should be used
+  when the value is meant to be precisely the specified number of bits and the
+  width restriction has a particular semantic meaning. For example, this could
+  be a timer value used within a system call interface, or a field in a TBF
+  header.
+- Machine-sized types (e.g., `usize`): These types should be used for a value
+  that should fit exactly into a register, like a abstract argument to the
+  `command` syscall.
```

Comment:
Is this example really what we want to have? Way back when we talked about this for the 2.0 system call interface, we converged on having the current system call interface and driver traits use `u32`s consistently, and implement an additional, 64-bit specific trait for such platforms when we have them. I know that at least for command args, we're currently not living this reality, though we arguably ought to be.

I still think this is a good idea, because this way we don't have to reason about edge-cases on different platforms for the system call interface. Driver (interfaces) behave identically across these platforms.

I'll have to check whether the CHERI PR (https://github.com/tock/tock/pull/4174) changes this.